### PR TITLE
set hostname using metadata for cloud-init

### PR DESCRIPTION
**Pull request by jacobweinstock**
This PR is update VM cloning for the vsphere provider to use cloud-init metadata to update the hostname of the VM.

### CODE REVIEW

File: pkg/provider/vsphere/cloudinit/metadata.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 // MetadataValues used
 type MetadataValues struct {
 	Hostname string
-	Networks []NetworkConfig
+	//Networks []NetworkConfig
```
Comment (j-griffith) [inline pkg/provider/vsphere/cloudinit/metadata.go:15]: Let's just remove this

File: pkg/provider/vsphere/cloudinit/metadata.go [hunk 2]
```diff
@@ -58,6 +58,33 @@ func GetMetadata(metadataValues *MetadataValues) ([]byte, error) {
...
+	}
+
+	return cloudinitMetaDataConfig, nil
+}
+
+const metadataTemplate = `
+instance-id: "{{ .Hostname }}"
+local-hostname: "{{ .Hostname }}"
+`
+
```
Comment (j-griffith) [inline pkg/provider/vsphere/cloudinit/metadata.go:87]: Debian 9, blah blah, Debian 9, for now we should be explicit that we're only supporting/working with Ubuntu LTS (18.04, but 20.04 soon).

File: pkg/provider/vsphere/cloudinit/metadata.go [hunk 3]
```diff
@@ -58,6 +58,33 @@ func GetMetadata(metadataValues *MetadataValues) ([]byte, error) {
...
+	metadata, err := GetMetadata(metadataValues)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get cloud metadata, %v", err)
+	}
+
+	var cloudinitMetaDataConfig Config
+
+	err = cloudinitMetaDataConfig.SetCloudInitMetadata(metadata)
+	if err != nil {
+		return nil, fmt.Errorf("unable to set cloud init metadata in extra config, %v", err)
```
Comment (bgentry-ntap) [inline pkg/provider/vsphere/cloudinit/metadata.go:77]: what is 'in extra config' here?

File: pkg/provider/vsphere/cloudinit/metadata.go [hunk 4]
```diff
@@ -58,6 +58,33 @@ func GetMetadata(metadataValues *MetadataValues) ([]byte, error) {
...
+// GenerateMetaData creates the meta data
+func GenerateMetaData(hostname string) (Config, error) {
+	// Create metadata
+	metadataValues := &MetadataValues{
+		Hostname: hostname,
+	}
+
+	metadata, err := GetMetadata(metadataValues)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get cloud metadata, %v", err)
```
Comment (bgentry-ntap) [inline pkg/provider/vsphere/cloudinit/metadata.go:70]: Does this work? It looks like we're returning nil for a struct value?

File: pkg/provider/vsphere/cloudinit/metadata.go (hunk 3 above)
Comment (jacobweinstock) [reply pkg/provider/vsphere/cloudinit/metadata.go]: bad wording. removed

File: pkg/provider/vsphere/cloudinit/metadata.go [hunk 5]
```diff
@@ -58,6 +58,33 @@ func GetMetadata(metadataValues *MetadataValues) ([]byte, error) {
 	return returnScript.Bytes(), nil
 }
 
+// GenerateMetaData creates the meta data
+func GenerateMetaData(hostname string) (Config, error) {
+	// Create metadata
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply pkg/provider/vsphere/cloudinit/metadata.go]: removed

File: pkg/provider/vsphere/cloudinit/metadata.go (hunk 2 above)
Comment (jacobweinstock) [reply pkg/provider/vsphere/cloudinit/metadata.go:74]: removed

File: pkg/provider/vsphere/cloudinit/metadata.go (hunk 1 above)
Comment (jacobweinstock) [reply pkg/provider/vsphere/cloudinit/metadata.go]: removed

File: pkg/provider/vsphere/cloudinit/metadata.go (hunk 4 above)
Comment (jacobweinstock) [reply pkg/provider/vsphere/cloudinit/metadata.go:57]: yeah, nil for the struct but only if there's an error.

Comment (bgentry-ntap) [reply pkg/provider/vsphere/cloudinit/metadata.go:57]: Ah, Config is a slice... not crazy about that name but not part of this PR

**Pull request merged**